src/python/pants/source/filespec.py

```diff
@@ -16,7 +16,7 @@ def glob_to_regex(pattern):
   :returns: A regex string that matches same paths as the input glob does.
   """
   out = ['^']
-  components = pattern.strip('/').replace('.', '[.]').split('/')
+  components = pattern.strip('/').replace('.', '[.]').replace('$', '[$]').split('/')
```

Comment:
At the time when this logic was introduced, pathspec was still using gitignore syntax which is different from glob matching syntax. I am not sure whether pathspec has evolved to support glob matching syntax since.